src/Libraries/Microsoft.Extensions.Telemetry.Testing/Metering/CollectedMeasurement.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.Telemetry.Testing.Metering;
+
+/// <summary>
+/// Represents a single measurement performed by an instrument.
+/// </summary>
+/// <typeparam name="T">The type of metric measurement value.</typeparam>
+[Experimental]
+public sealed class CollectedMeasurement<T>
+    where T : struct
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CollectedMeasurement{T}"/> class.
+    /// </summary>
+    /// <param name="value">The measurement's value.</param>
+    /// <param name="tags">The dimensions of this measurement.</param>
+    /// <param name="timestamp">The time that the measurement occurred at.</param>
+    internal CollectedMeasurement(T value, ReadOnlySpan<KeyValuePair<string, object?>> tags, DateTimeOffset timestamp)
+    {
+        var d = new Dictionary<string, object?>();
+        foreach (var tag in tags)
+        {
+            d[tag.Key] = tag.Value;
+        }
+
+        Tags = d;
+        Timestamp = timestamp;
+        Value = value;
+    }
+
+    /// <summary>
+    /// Gets a measurement's value.
+    /// </summary>
+    public T Value { get; }
+
+    /// <summary>
+    /// Gets a timestamp indicating when the measurement was recorded.
+    /// </summary>
+    public DateTimeOffset Timestamp { get; }
+
+    /// <summary>
+    /// Gets the measurement's dimensions.
+    /// </summary>
+    public IReadOnlyDictionary<string, object?> Tags { get; }
+
+    /// <summary>
+    /// Gets a string representing the measurement.
+    /// </summary>
+    /// <returns>The string representation of this measurement.</returns>
+    public override string ToString() => $"[{Timestamp.ToString("HH:mm:ss.ffff", CultureInfo.InvariantCulture)}] {Value}";
```

Comment:
nit: I think this would be better in a DebuggerDisplay attribute. I don't imagine anyone would use this string at runtime.

Reply:
Can I express this as an attribute? As it is, the debugger does call ToString so its shows up in the UI. I'd be happy with it as a display attribute instead if that can be expressed.

Reply:
Yes. See https://github.com/dotnet/runtime/pull/87754 for some examples.

Add:

```csharp
internal string DebuggerToString()
{
    return $"[{Timestamp.ToString("HH:mm:ss.ffff", CultureInfo.InvariantCulture)}] {Value}";
}
```

and

```csharp
[DebuggerDisplay("{DebuggerToString(),nq}")]
```